File: android/app/src/main/kotlin/dev/omnesis/android/ui/agent/AgentWatchCard.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.agent

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bolt
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import dev.omnesis.android.designsystem.theme.OmRadius
import dev.omnesis.android.designsystem.theme.OmTheme
import dev.omnesis.android.transport.dto.AgentToolResult

/**
 * Tool names whose successful call is an automation write worth a card.
 * `watch_create` / `watch_update` are what the agent calls today; the retired
 * `trigger_*` authoring names stay listed so transcripts recorded before
 * watches still render their cards when reopened.
 */
val AGENT_AUTOMATION_TOOLS = setOf("watch_create", "watch_update", "trigger_upsert", "trigger_toggle")

/**
 * Inline transcript card the agent surfaces whenever it successfully writes an
 * automation. Mirrors iOS `AgentWatchCard`:
 *
 *   - **Lightning glyph** (`bolt.fill`) in `accent` on the leading edge — the
 *     consistent "agent just performed an action" cue.
 *   - **Primary line**: verb + automation name. Verb derives from the result
 *     kind: created / updated for `trigger.upserted`, and enabled / disabled
 *     for a `trigger.toggled` result in an older transcript.
 *   - **Secondary line** (`triggerUpserted` only): the agent-supplied one-line
 *     summary explaining what the automation does.
 *   - **Pending stub**: while the tool block is open but the result hasn't landed,
 *     a muted "Setting up watch…" line so the user sees something is in flight.
 *   - **Error**: delegates to [AgentToolErrorCard].
 *
 * Read-only trigger tools (`triggers_list`, `trigger_get`, `trigger_firings`)
 * intentionally do NOT route here — they're background data fetches, not
 * user-visible actions.
 *
 * The card reports what the agent did and is not tappable: this app has no
 * watch surface to open, so a tap would have nowhere to land. The watch is
 * reachable from the portal and from the phone that has a Watches section.
 */
@Composable
fun AgentWatchCard(call: AgentToolCall) {
    when (val resolved = resolveWatch(call)) {
        is WatchResolved.Pending -> WatchPendingBody()

        is WatchResolved.Error ->
            AgentToolErrorCard(code = resolved.code, message = resolved.message)

        is WatchResolved.Action ->
            WatchActionBody(
                verb = resolved.verb,
                name = resolved.name,
                summary = resolved.summary,
            )
    }
}

/**
 * Pending stub — a neutral `border` left rail, muted bolt + "Setting up watch…".
 * No fill, no clip (transparent background bleeds the transcript through). Matches
 * iOS `pendingBody`.
 */
@Composable
private fun WatchPendingBody() {
    Row(
        Modifier.fillMaxWidth().height(IntrinsicSize.Min),
    ) {
        Box(Modifier.width(2.dp).fillMaxHeight().background(OmTheme.colors.border))
        Row(
            Modifier.padding(horizontal = 8.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Icon(
                Icons.Filled.Bolt,
                contentDescription = null,
                tint = OmTheme.colors.textMuted,
                modifier = Modifier.size(13.dp),
            )
            Text(
                "Setting up watch…",
                style = MaterialTheme.typography.bodySmall,
                color = OmTheme.colors.textMuted,
            )
        }
    }
}

/**
 * Action card — a solid `accent` 2-dp left rail, `accent @ 0.06` fill clipped to a
 * 6-dp rounded rect, bolt + verb (`textPrimary`) + name (`accent`, single line) and
 * an optional summary line (`textSecondary`, up to 2 lines). The whole card is
 * tappable. Matches iOS `actionBody`.
 */
@Composable
private fun WatchActionBody(
    verb: String,
    name: String,
    summary: String?,
) {
    Row(
        Modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
            .clip(RoundedCornerShape(OmRadius.medium))
            .background(OmTheme.colors.accent.copy(alpha = 0.06f)),
    ) {
        Box(Modifier.width(2.dp).fillMaxHeight().background(OmTheme.colors.accent))
        Row(
            Modifier.padding(horizontal = 8.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.Top,
        ) {
            Icon(
                Icons.Filled.Bolt,
                contentDescription = null,
                tint = OmTheme.colors.accent,
                modifier = Modifier.size(13.dp).padding(top = 2.dp),
            )
            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    Text(
                        verb,
                        style = MaterialTheme.typography.bodySmall,
                        fontWeight = FontWeight.Medium,
                        color = OmTheme.colors.textPrimary,
                    )
                    Text(
                        name,
                        style = MaterialTheme.typography.bodySmall,
                        fontWeight = FontWeight.Medium,
                        color = OmTheme.colors.accent,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                if (!summary.isNullOrEmpty()) {
                    Text(
                        summary,
                        style = MaterialTheme.typography.labelMedium,
                        color = OmTheme.colors.textSecondary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
        }
    }
}

/**
 * Compact, expandable error card. Same visual family as [AgentToolResultErrorView]
 * but tuned for the trigger surface: a fainter `danger @ 0.04` fill, the first line
 * rendered in `textSecondary` (not danger), and a `textMuted` chevron. Lifted into a
 * standalone composable so the trigger card can render tool errors with one look.
 * Mirrors iOS `AgentToolErrorCard`.
 */
@Composable
fun AgentToolErrorCard(code: String, message: String) {
    var expanded by remember { mutableStateOf(false) }
    val first = message.substringBefore('\n')
    val multi = message != first

    Row(
        Modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
            .clip(RoundedCornerShape(OmRadius.small))
            .background(OmTheme.colors.danger.copy(alpha = 0.04f)),
    ) {
        Box(Modifier.width(2.dp).fillMaxHeight().background(OmTheme.colors.danger.copy(alpha = 0.45f)))
        Column(
            Modifier
                .then(if (multi) Modifier.clickable { expanded = !expanded } else Modifier)
                .padding(horizontal = 8.dp, vertical = 4.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Text(
                    code,
                    style = MaterialTheme.typography.labelSmall.copy(fontFamily = FontFamily.Monospace),
                    fontWeight = FontWeight.SemiBold,
                    color = OmTheme.colors.danger,
                )
                Text(
                    first,
                    style = MaterialTheme.typography.labelMedium,
                    color = OmTheme.colors.textSecondary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                if (multi) {
                    Icon(
                        if (expanded) Icons.Filled.KeyboardArrowUp else Icons.Filled.KeyboardArrowDown,
                        contentDescription = null,
                        tint = OmTheme.colors.textMuted,
                        modifier = Modifier.size(13.dp),
                    )
                }
            }
            AnimatedVisibility(visible = expanded && multi) {
                Text(
                    message,
                    style = MaterialTheme.typography.labelSmall.copy(fontFamily = FontFamily.Monospace),
                    color = OmTheme.colors.textSecondary,
                )
            }
        }
    }
}

// MARK: - Result interpretation

private sealed interface WatchResolved {
    data object Pending : WatchResolved
    data class Error(val code: String, val message: String) : WatchResolved
    data class Action(
        val verb: String,
        val name: String,
        val summary: String?,
        val watchId: String,
    ) : WatchResolved
}

/**
 * Maps a watch tool call's result kind to the rendered verb + payload. A missing
 * result is "pending"; an error result delegates to the error card; the upserted
 * and toggled kinds map to action verbs. Any other kind that somehow reaches
 * here (routing is gated upstream) collapses to "pending" so nothing crashes if
 * the contract drifts. Mirrors iOS `AgentWatchCard.resolve()`.
 */
private fun resolveWatch(call: AgentToolCall): WatchResolved =
    when (val result = call.result) {
        null -> WatchResolved.Pending

        is AgentToolResult.ErrorResult ->
            WatchResolved.Error(code = result.code, message = result.message)

        is AgentToolResult.WatchUpserted ->
            WatchResolved.Action(
                verb = if (result.action == "created") "Created watch" else "Updated watch",
                name = result.name,
                summary = result.summary,
                watchId = result.watchId,
            )

        // Cards from conversations stored before the rename. Rendered rather
        // than dropped: they still describe something the user asked for, and
        // a stored transcript is still theirs.
        is AgentToolResult.TriggerUpserted ->
            WatchResolved.Action(
                verb = if (result.action == "created") "Created watch" else "Updated watch",
                name = result.name,
                summary = result.summary,
                watchId = result.triggerId,
            )

        is AgentToolResult.TriggerToggled ->
            WatchResolved.Action(
                verb = if (result.enabled) "Enabled watch" else "Disabled watch",
                name = result.name,
                summary = null,
                watchId = result.triggerId,
            )

        else -> WatchResolved.Pending
    }

File: ios/Sources/Omnesis/UI/Agent/AgentWatchCard.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

/// Tool names whose successful call is an automation write worth a card.
/// `watch_create` / `watch_update` are what the agent calls today; the retired
/// `trigger_*` authoring names stay listed so transcripts recorded before
/// watches still render their cards when reopened.
let agentAutomationTools: Set<String> = [
    "watch_create", "watch_update", "trigger_upsert", "trigger_toggle",
]

/// Whether an automation card handles this tool call.
func agentAutomationCardHandles(_ tool: String) -> Bool {
    agentAutomationTools.contains(tool)
}

/// Inline transcript card the agent surfaces every time it successfully
/// writes an automation. Mirrors the portal's `TriggerActionCard`:
///
///   - **Lightning glyph** (SF Symbol `bolt.fill`) in `Theme.accent` on
///     the leading edge — the consistent "agent just performed an
///     action" cue.
///   - **Primary line**: verb + automation name. Verb is derived from the
///     result kind: created / updated for `trigger.upserted`, and
///     enabled / disabled for a `trigger.toggled` result in an older
///     transcript.
///   - **Secondary line** (`triggerUpserted` only): the one-line summary
///     explaining what the automation does.
///   - **Pending stub**: while the tool block is open but the result
///     hasn't landed yet, show a muted "Setting up watch…" line so
///     the user sees something is in flight.
///
/// Read-only trigger tools (`triggers_list`, `trigger_get`,
/// `trigger_firings`) intentionally do NOT route here — they're
/// background data fetches, not user-visible actions.
@available(iOS 17.0, *)
struct AgentWatchCard: View {
    @Environment(NotificationRouter.self) private var router
    let call: AgentToolCall

    var body: some View {
        switch resolve() {
        case .pending:
            pendingBody
        case .error(let code, let message):
            AgentToolErrorCard(code: code, message: message)
        case .action(let verb, let name, let summary, let watchId):
            Button {
                // Channel the tap through the existing NotificationRouter so
                // it shares one deep-link path with a watch-firing push. The
                // HomeView flips to the Watches section and WatchesView's
                // consumeRouterTarget pushes the watch onto the nav stack —
                // where the person can read what it will do and stop it if
                // that is not what they meant.
                router.pendingTarget = .watch(watchId: watchId)
            } label: {
                actionBody(verb: verb, name: name, summary: summary)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Text("\(verb) \(name) — open watch details"))
        }
    }

    private var pendingBody: some View {
        HStack(alignment: .firstTextBaseline, spacing: 8) {
            Image(systemName: "bolt.fill")
                .font(.system(size: 11, weight: .medium))
                .foregroundStyle(Theme.textMuted)
            Text("Setting up watch…")
                .font(.system(size: 13))
                .foregroundStyle(Theme.textMuted)
            Spacer(minLength: 0)
        }
        .padding(.vertical, 6)
        .padding(.horizontal, 8)
        .background(
            HStack {
                Rectangle()
                    .fill(Theme.border)
                    .frame(width: 2)
                Spacer()
            }
        )
    }

    private func actionBody(verb: String, name: String, summary: String?) -> some View {
        HStack(alignment: .top, spacing: 8) {
            Image(systemName: "bolt.fill")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(Theme.accent)
                .padding(.top, 2)
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 4) {
                    Text(verb)
                        .font(.system(size: 13, weight: .medium))
                        .foregroundStyle(Theme.textPrimary)
                    Text(name)
                        .font(.system(size: 13, weight: .medium))
                        .foregroundStyle(Theme.accent)
                        .lineLimit(1)
                        .truncationMode(.tail)
                }
                if let summary, !summary.isEmpty {
                    Text(summary)
                        .font(.system(size: 12))
                        .foregroundStyle(Theme.textSecondary)
                        .lineLimit(2)
                }
            }
            Spacer(minLength: 0)
        }
        .padding(.vertical, 8)
        .padding(.horizontal, 8)
        .background(
            HStack {
                Rectangle()
                    .fill(Theme.accent)
                    .frame(width: 2)
                Color.clear
            }
        )
        .background(Theme.accent.opacity(0.06))
        .clipShape(RoundedRectangle(cornerRadius: 6, style: .continuous))
    }

    // MARK: - Result interpretation

    private enum Resolved {
        case pending
        case error(code: String, message: String)
        case action(verb: String, name: String, summary: String?, watchId: String)
    }

    private func resolve() -> Resolved {
        guard let result = call.result else { return .pending }
        switch result {
        case .error(let code, let message):
            return .error(code: code, message: message)
        case .watchUpserted(let watchId, let name, let action, _, let summary):
            let verb = action == "created" ? "Created watch" : "Updated watch"
            return .action(verb: verb, name: name, summary: summary, watchId: watchId)
        case .triggerUpserted(let watchId, let name, let action, _, let summary):
            // Cards from conversations stored before the rename. Rendered
            // rather than dropped: they still describe something the user
            // asked for, and a stored transcript is still theirs.
            let verb = action == "created" ? "Created watch" : "Updated watch"
            return .action(verb: verb, name: name, summary: summary, watchId: watchId)
        case .triggerToggled(let watchId, let name, let enabled):
            return .action(
                verb: enabled ? "Enabled watch" : "Disabled watch",
                name: name,
                summary: nil,
                watchId: watchId
            )
        default:
            // Any other result that somehow reaches this card (routing is
            // gated upstream in AgentPartView) collapses to a "pending"
            // placeholder so nothing crashes if the contract ever drifts.
            return .pending
        }
    }
}

// MARK: - Reusable error card (lifts out of AgentToolCallView for AgentWatchCard reuse)

/// Compact, expandable error card. Identical visual language to
/// `AgentToolCallView`'s inline error rendering; lifted into a
/// standalone view so other transcript surfaces (like
/// `AgentWatchCard`) can render tool errors with the same look.
@available(iOS 17.0, *)
struct AgentToolErrorCard: View {
    let code: String
    let message: String

    @State private var expanded: Bool = false

    var body: some View {
        let first = firstLine(message)
        let hasMore = message != first
        VStack(alignment: .leading, spacing: 4) {
            Button {
                if hasMore {
                    withAnimation(.easeInOut(duration: 0.15)) { expanded.toggle() }
                }
            } label: {
                HStack(spacing: 6) {
                    Text(code)
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(Theme.danger)
                    Text(first)
                        .font(.system(size: 12))
                        .foregroundStyle(Theme.textSecondary)
                        .lineLimit(1)
                        .truncationMode(.tail)
                    Spacer(minLength: 0)
                    if hasMore {
                        Image(systemName: expanded ? "chevron.up" : "chevron.down")
                            .font(.system(size: 9, weight: .medium))
                            .foregroundStyle(Theme.textMuted)
                    }
                }
            }
            .buttonStyle(.plain)
            if expanded, hasMore {
                Text(message)
                    .font(Theme.monospace(size: 11))
                    .foregroundStyle(Theme.textSecondary)
                    .textSelection(.enabled)
            }
        }
        .padding(.vertical, 4)
        .padding(.horizontal, 8)
        .background(
            HStack {
                Rectangle()
                    .fill(Theme.danger.opacity(0.45))
                    .frame(width: 2)
                Color.clear
            }
        )
        .background(Theme.danger.opacity(0.04))
        .clipShape(RoundedRectangle(cornerRadius: 4, style: .continuous))
    }

    private func firstLine(_ text: String) -> String {
        guard let idx = text.firstIndex(of: "\n") else { return text }
        return String(text[..<idx])
    }
}

#if DEBUG
@available(iOS 17.0, *)
#Preview("Trigger card — created") {
    let call = AgentToolCall(
        toolCallId: "tu_1",
        tool: "trigger_upsert",
        args: JSONAny(value: NSNull()),
        argsSummary: "create",
        argsKnown: true,
        result: .watchUpserted(
            watchId: "wat_a",
            name: "Email digest",
            action: "created",
            enabled: true,
            summary: "Notify when a new invoice email arrives in Gmail"
        ),
        durationMs: 120
    )
    return VStack(alignment: .leading) {
        AgentWatchCard(call: call)
    }
    .padding()
    .background(Theme.bgPrimary)
    .environment(NotificationRouter())
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("Trigger card — updated, no summary") {
    let call = AgentToolCall(
        toolCallId: "tu_2",
        tool: "trigger_upsert",
        args: JSONAny(value: NSNull()),
        argsSummary: "update",
        argsKnown: true,
        result: .watchUpserted(
            watchId: "wat_b",
            name: "Daily summary",
            action: "updated",
            enabled: true,
            summary: nil
        ),
        durationMs: 95
    )
    return VStack(alignment: .leading) {
        AgentWatchCard(call: call)
    }
    .padding()
    .background(Theme.bgPrimary)
    .environment(NotificationRouter())
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("Watch card — toggled off") {
    let call = AgentToolCall(
        toolCallId: "tu_3",
        tool: "trigger_toggle",
        args: JSONAny(value: NSNull()),
        argsSummary: "disable",
        argsKnown: true,
        result: .triggerToggled(
            triggerId: "trg_c",
            name: "Workout reminder",
            enabled: false
        ),
        durationMs: 40
    )
    return VStack(alignment: .leading) {
        AgentWatchCard(call: call)
    }
    .padding()
    .background(Theme.bgPrimary)
    .environment(NotificationRouter())
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("Trigger card — pending (no result yet)") {
    let call = AgentToolCall(
        toolCallId: "tu_4",
        tool: "trigger_upsert",
        args: JSONAny(value: NSNull()),
        argsSummary: "create",
        argsKnown: true,
        result: nil,
        durationMs: nil
    )
    return VStack(alignment: .leading) {
        AgentWatchCard(call: call)
    }
    .padding()
    .background(Theme.bgPrimary)
    .environment(NotificationRouter())
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("Trigger card — error") {
    let call = AgentToolCall(
        toolCallId: "tu_5",
        tool: "trigger_upsert",
        args: JSONAny(value: NSNull()),
        argsSummary: "create",
        argsKnown: true,
        result: .error(
            code: "non_ios_actions",
            message: "every action in a trigger the agent manages must be `notify-ios`; got exec"
        ),
        durationMs: 8
    )
    return VStack(alignment: .leading) {
        AgentWatchCard(call: call)
    }
    .padding()
    .background(Theme.bgPrimary)
    .environment(NotificationRouter())
    .preferredColorScheme(.dark)
}
#endif
#endif
